### ISSUE TITLE
Proposal: deprecate some logging options.

### ISSUE THREAD

**Issue by howardjohn** [istio/istio#46012]
we currently have the following options:

```
	stringArrayVar(&o.OutputPaths, "log_target", o.OutputPaths,
		"The set of paths where to output the log. This can be any path as well as the special values stdout and stderr")

	stringVar(&o.RotateOutputPath, "log_rotate", o.RotateOutputPath,
		"The path for the optional rotating log file")

	intVar(&o.RotationMaxAge, "log_rotate_max_age", o.RotationMaxAge,
		"The maximum age in days of a log file beyond which the file is rotated (0 indicates no limit)")

	intVar(&o.RotationMaxSize, "log_rotate_max_size", o.RotationMaxSize,
		"The maximum size in megabytes of a log file beyond which the file is rotated")

	intVar(&o.RotationMaxBackups, "log_rotate_max_backups", o.RotationMaxBackups,
		"The maximum number of log file backups to keep before older files are deleted (0 indicates no limit)")

	boolVar(&o.JSONEncoding, "log_as_json", o.JSONEncoding,
		"Whether to format output as JSON or in plain console-friendly format")

	stringVar(&o.outputLevels, "log_output_level", o.outputLevels,
			fmt.Sprintf("Comma-separated minimum per-scope logging level of messages to output, in the form of "+
				"<scope>:<level>,<scope>:<level>,... where scope can be one of [%s] and level can be one of %s",
				s, levelListString))

		stringVar(&o.stackTraceLevels, "log_stacktrace_level", o.stackTraceLevels,
			fmt.Sprintf("Comma-separated minimum per-scope logging level at which stack traces are captured, in the form of "+
				"<scope>:<level>,<scope:level>,... where scope can be one of [%s] and level can be one of %s",
				s, levelListString))

		stringVar(&o.logCallers, "log_caller", o.logCallers,
			fmt.Sprintf("Comma-separated list of scopes for which to include caller information, scopes can be any of [%s]", s))
```

Standard usage and best practice is to log to stdout. I don't think we are, or should be in, the business, of log rotation. Anyone that needs this can pipe to a much better suited application to do this.

Therefor I recommend:

**Definitely deprecate** `log_rotate`, `log_rotate_max_age`, `log_rotate_max_size`, and `log_rotate_max_backups`.

**Probably deprecate** `log_target` - I am less sure of since being able to log to stderr is a reasonable use case.

**Must keep** `--log_output_level`, `--log_as_json`

**Probably keep** `log_stacktrace_level`, `log_caller`

This will make https://github.com/istio/istio/issues/44683 easier. Without this, we will need to re-implement these for slog.

**Comment by howardjohn** [comment]
cc @costinm @zirain @ramaraochavali @stevenctl - anyone concerned about these flags going away?

**Comment by keithmattix** [comment]
![image](https://github.com/istio/istio/assets/1531662/92b42f41-5248-436c-9d0c-743a288ce1fa)

**Comment by costinm** [comment]
I would wait few weeks on this one - until golang 1.21 is out with the new slog package. Like OTel for metrics, I think the standard core logging API is better than every project inventing its own. 

I'm all for deprecating the one you propose - but I would go deeper: we need to switch to OTel ( metrics - but also traces and logs ), and OTel has its own environment variables for settings some of this. 

Since we have the backward compat and upgrade issue lingering - we can just leave the current logging as is - and fully deprecate it ( but working for a few releases ). No need to worry about some user needing those specific flags. 
At the same time - get the new slog and otel in.

I looked a bit - and it is possible to write a custom slog Handler that logs with our existing package, so we can convert all our internal use to slog and keep the logger as balast for few releases. 

What I am in doubt about is the relation between slog and OTel logs/traces - I think Istiod support can be greatly improved
if we start using traces and the tooling around it ( for pushes, connection) - with the sampling common to tracers - while some info would fit the 'access log'/audit pattern. And with the recent convergence of Elastic Common Schema with OTel semantic conventions - adopting the common patterns in Istiod (and ztunnel, CNI) would also help a lot operations. 

In other words: normally I would be very enthusiastic about this PR, but in current context I would rather have a plan to 
replace the entire log package - with a (long?) transition period where it can still be used.

**Comment by costinm** [comment]
In particular for those option - if I remember correctly they are mostly for VM usage - the agent but also  Istiod can run in a VM, and our stderr didn't look well with journald.

**Comment by howardjohn** [comment]
Ack, sounds like we have a similar end goal but different thoughts on getting there. I was planning to write up a design on our holistic control plane observability strategy (metrics, logs, and tracing) but split this out. Maybe worth holding this and including it in the doc, where I can also look into everything you mentioned in https://github.com/istio/istio/issues/46012#issuecomment-1635966616.

Re VMs - we already just log to stdout in our istio-start.sh, and if users want to do it better/properly they can set `StandardOut=journal` (https://www.freedesktop.org/software/systemd/man/systemd.exec.html#StandardOutput=) I think? Not sure though, minimal experience with systemd

**Comment by costinm** [comment]
Agreed - and there are plenty of options on VMs, in particular if we start using OTel protocol - users can just run a collector and let it handle it. 

I think one way or another, sooner or later we will end up in the same place, using open telemetry. We may get there faster or slower - and migrations will complicate things, I think that's where we need to be extra careful.

The issue is if we should start making breaking changes gradually - and deprecate things before the replacement is ready, or add a second path with OTel and an option to use it. With the wrappers we have in place - the second option 
is quite possible.

**Comment by howardjohn** [comment]
do you expect otel for logs as well? I was assuming only stdout (from slog)
long term - otel collector (and everything else out there) can pull from
that?

For traces and metrics of course

**Comment by costinm** [comment]
On Fri, Jul 14, 2023 at 8:36 PM John Howard ***@***.***>
wrote:

> do you expect otel for logs as well? I was assuming only stdout (from
> slog)
> long term - otel collector (and everything else out there) can pull from
> that?
>
> For traces and metrics of course



Why not ?

I see most logs as events that are part of traces. Most of the control
plane logs are associated with
startup ( which should probably generate a trace) or even push or config
distributions.

I don't mind a slog + traces/metrics approach - keeping logs in slog, but
given that slog supports
custom Handler I don't see any reason to not have an option for slog to
send the logs using the OTLP,
either as part of traces or as the (least mature but still good enough)
logging spec.

As API, we can use slog - it is relatively easy to swap the wrapper we use
with slog, but the key would be to
add metadata - and if we can associate the log with one of the traceable
categories, I think it is possible to
have Handler that attaches the log to the current trace.

I just did a bit of prototyping, and didn't try it end to end, but I
suspect we're both discovering the
power of tracing :-), and having a unified telemetry story is quite
interesting.

One piece we need to preserve is the flexibility to dynamically configure
log levels - and in the 'unified'
view we could extend it to also configure different tracing categories (
i.e. a trace can be treated as a log
with a specific name). I am not sure if dynamic using the current http
interface - or dynamic using our own
telemetry API, but OTel is lacking on dynamic config ( AFAIK ).

**Comment by howardjohn** [comment]
Agree with everything you said.

Especially the "trace category/levels" - I don't know if there is a
standard there but it seems pretty nice. I am all in on debugging via
traces now - but can't afford 100% sampling of all of Istio all the time :-)

Rust `tracing` library (which is basically a logging library) has levels
matching log levels (again, it's literally a logging library :-) ) so there
is some prior art there at least, if not in Go/core otel.

**Comment by hanxiaop** [comment]
not stale

**Comment by hanxiaop** [comment]
not stale
